### ISSUE TITLE
Closures

### ISSUE THREAD

**Issue by lcd047** [lavv17/lftp#360]
Segmented and parallel transfers would benefit a lot from a more flexible mechanism of closures.  For example these don't work:
```
set net:limit-total-rate/some.host              10M
set mirror:use-pget-n/some.host                 5
set mirror:parallel-directories/some.host       yes
set mirror:parallel-transfer-count/some.host    5
```
(the point here being to limit the total load inflicted to a given site when using parallel connections, as opposed to setting global limits, or limits for individual connections).

Can these please be considered for some future version?

**Comment by lavv17** [comment]
Good idea, but remember that mirror can transfer between two remote locations.

**Comment by lcd047** [comment]
Transferring between two remote locations is a misfeature of the FTP protocol, yes.  (Does anybody still use that feature?  It brings back memories from 30+ years ago. :smile:)  Anyway, if `mirror` specifies two locations with conflicting limits I suppose the more restrictive / conservative ones should be chosen.

**Comment by lavv17** [comment]
Ok, seems feasible to do.
BTW, FXP is still in use, I have a few reports. But largely it is
prohibited as it poses a security risk. But mirror can be used to transfer
between two sites without FXP, even between two different protocols. No
such a cute optimization of traffic though.

вс, 9 июля 2017, 13:02 LCD 047 <notifications@github.com>:

**Comment by lavv17** [comment]
I've implemented limit-total-rate with closures to limit on per-host basis: https://github.com/lavv17/lftp/tree/per-host-limit-total-rate

**Comment by lcd047** [comment]
Thank you.

I get an error when compiling lftp with gcc 7.1:
```
In file included from ResMgr.cc:26:0:
../lib/math.h: In function ‘int signbit(float)’:
../lib/math.h:2576:1: error: ‘int signbit(float)’ conflicts with a previous declaration
 _GL_MATH_CXX_REAL_FLOATING_DECL_2 (signbit)
 ^
In file included from /usr/include/c++/7.1.0/math.h:36:0,
                 from ../lib/math.h:27,
                 from ResMgr.cc:26:
/usr/include/c++/7.1.0/cmath:668:3: note: previous declaration ‘constexpr bool std::signbit(float)’
   signbit(float __x)
   ^~~~~~~
```
After removing the offending lines from `lib/math.h` it seems to work fine, but the right fix is probably more involved.

**Comment by lavv17** [comment]
math.h problem seem to be gnulib related. It can be fixed by including cmath, but I'm not sure what problems gnulib's math.h is striving to fix.

**Comment by lavv17** [comment]
closures added in a654835817015f5707b7868d94144b7cf3e63a24 and per-host limit merged in c279613a5b619d6ce72c0c2d08cd09d69a024656
